### ISSUE TITLE
DataParallel doesn't seem to work

### ISSUE THREAD

**Issue by George3d6** [huggingface/accelerate#124]
I'm using the library in what I though would be a fairly bog-standard way:

```
             accelerator = Accelerator()
             model, optimizer, train_dl = accelerator.prepare(self.model, optimizer, train_dl)
             for _ in range(blah):
                    error = run_epoch(model, train_dl, criterion, optimizer, scaler, accelerator)

            ....
            # inside `run_epoch` I do back prop using:
            accelerator.backward(scaler.scale(loss))
```

The reason I'm using a separate scaler (+ separate autocasting) is because that functionality doesn't seem to work, and it's unclear from the doc whehter it's supposed to work.

However doing thing this way results in no parallelism.

On a g4dn AWS instance where GPU training works with:
`NVIDIA-SMI 470.42.01    Driver Version: 470.42.01    CUDA Version: 11.4` and 4x GPUs

I only see 1x GPU working according to `nvidia-smi` (as in, the other are using no memory and are at 0% CPU)

How do I go about debugging this? Might there be driver issues that could cause this problem?

**Comment by sgugger** [comment]
Mixed precision will work as long as you set `fp16` to `True` in your accelerate config.

How are you launching the script? If it is with `accelerate launch`, what is the content of your config file?

**Comment by George3d6** [comment]
> Mixed precision will work as long as you set fp16 to True in your accelerate config.

I tried setting that but I got an error about using `Float` instead of `Half` in the `backwards` function (after havving had already removed all of my scaling + autocasting)

Is there any issue if I just use my own scaler + autocasting to half during training?

> Mixed precision will work as long as you set fp16 to True in your accelerate config.

I missed that bit and was just launching normally using an interpreter.

But launching with `accelerate launch` instead of `python3` still results in the same behaviour.

**Comment by George3d6** [comment]
by "same behaviour" I mean "only one GPU is getting used"

**Comment by George3d6** [comment]
Update: Behavior remains the same if I configure as:

```
In which compute environment are you running? ([0] This machine, [1] AWS (Amazon SageMaker)): 0
Which type of machine are you using? ([0] No distributed training, [1] multi-GPU, [2] TPU): 1
How many different machines will you use (use more than 1 for multi-node training)? [1]: 1
How many processes in total will you use? [1]: 1
Do you wish to use FP16 (mixed precision)? [yes/NO]: yes
```

**Comment by sgugger** [comment]
If you launch the script with python it will indeed use only one GPU, you have to use the PyTorch distributed launcher or `accelerate launch` to properly launch you script in  distributed fashion. For PyTorch distributed launcher, look at the PyTorch documentation or  [this README](https://github.com/huggingface/accelerate/tree/main/examples). For the `accelerate launch`, you need to execute `accelerate config` beforehand.

If you tell `accelerate config` you only want to use 1 process and not 4, the script will be very obedient and only use 1 GPU and not 4 ;-) .

**Comment by George3d6** [comment]
Got in, I guess it might be worth renaming process to "devices" then? I took it to mean something like "are you launching processes with torch.mp", since I had already selected "multiple gpus"... but maybe my interpretation was just dumb.

Anyway, it seems to kinda work now in that I get a:
`RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:3! (when checking arugment for argument index in method wrapper_index_select` error (not sure why, but it happens outside the training loop, so maybe I'm not unwrapping properly.

At any rate, it seems that launching this way actually acceses accelerate to, ahm, launch the script I'm asking it to run 4x time :))

But again, maybe it's a usage bug somewhere. Anyway, the original issue was just misreading the docuemtnation, so closing this, sorry for the bother, and good luck with the project :) In principle I think the idea is great
